sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala

```diff
@@ -294,4 +295,23 @@ class SparkSessionExtensions {
   def injectTableFunction(functionDescription: TableFunctionDescription): Unit = {
     injectedTableFunctions += functionDescription
   }
+
+  private[this] val runtimeOptimizerRules = mutable.Buffer.empty[RuleBuilder]
+
+  private[sql] def buildRuntimeOptimizerRules(session: SparkSession): Seq[Rule[LogicalPlan]] = {
+    runtimeOptimizerRules.map(_.apply(session)).toSeq
+  }
+
+  /**
+   * Inject a runtime `Rule` builder into the [[SparkSession]].
+   * The injected rules will be executed after built-in
+   * [[org.apache.spark.sql.execution.adaptive.AQEOptimizer]] rules are applied.
+   * A runtime optimizer rule is used to improve the quality of a logical plan during execution
+   * which can leverage accurate statistics from shuffle.
+   *
+   * Note that, it does not work if adaptive query execution is disabled.
+   */
+  def injectRuntimeOptimizerRule(builder: RuleBuilder): Unit = {
+    runtimeOptimizerRules += builder
+  }
```

Comment:
Shall we move these new additions somewhere around the exiting `queryStagePrepRuleBuilders` and `buildQueryStagePrepRules`?